### PR TITLE
JWT & passport authentication

### PR DESCRIPTION
**Pull request by jsshotts**

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -2,25 +2,29 @@
...
-  "main": "index.js",
+  "main": "server.js",
   "scripts": {
     "start": "nodemon server.js",
     "test": "jest"
   },
   "author": "Team 805",
   "license": "ISC",
   "dependencies": {
+    "-": "^0.0.1",
```
Comment (adrien-chaussabel) [inline backend/package.json]: I think this might be mistakenly imported package

File: backend/package.json [hunk 2]
```diff
@@ -2,25 +2,29 @@
...
     "dotenv": "^8.2.0",
     "express": "^4.17.1",
     "express-session": "^1.17.1",
+    "jsonwebtoken": "^8.5.1",
     "mongoose": "^5.11.15",
     "passport": "^0.4.1",
     "passport-jwt": "^4.0.0",
-    "passport-local": "^1.0.0"
+    "passport-local": "^1.0.0",
+    "save-dev": "^0.0.1-security"
```
Comment (adrien-chaussabel) [inline backend/package.json]: Ditto here about importing by mistake

File: frontend/cypress/integration/tests/auth_spec.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+const backend = 'localhost:3001';
+const testUsername = 'test';
+const testPassword = 'test';
+// Must be running the auth server for all tests.
+
+describe('Login test', () => {
```
Comment (adrien-chaussabel) [inline frontend/cypress/integration/tests/auth_spec.js]: Does this need to be a cypress test? Can it be done in Jest?

File: frontend/src/Login/Login.js [hunk 4]
```diff
@@ -6,6 +6,26 @@ import Button from '@material-ui/core/Button';
 import axios from 'axios';
 import './Login.css';
 
+export const login = (username, password) => {
+  if (!username || !password) {
+    alert('username and password required!');
```
Comment (adrien-chaussabel) [inline frontend/src/Login/Login.js:11]: For now this is fine since we haven't worked on error messaging, but generally it is advised to not use alert().

File: frontend/cypress/integration/tests/auth_spec.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  it('Logs into cloudhaven, then refreshes token', function() {
+    // hard coded login for now
+    cy.request('POST', backend + '/login', 
+      { username: testUsername, password: testPassword})
+      .then((response) => {
+        // we can't access the tokens to test they are different.
+        cy.request('POST', backend + '/token');
+      })
+  });
+});
```
Comment (mlkrajewski) [inline frontend/cypress/integration/tests/auth_spec.js]: tests look pretty thorough

File: frontend/src/Login/Login.js [hunk 6]
```diff
@@ -76,7 +67,7 @@ const Login = () => {
           fullWidth
           variant="contained"
           color="primary"
-          onClick={login}>
+          onClick={() => login(loginUsername, loginPassword)}>
```
Comment (adrien-chaussabel) [inline frontend/src/Login/Login.js:70]: Don't need to pass in loginUsername and password since they are state variables

File: frontend/cypress/integration/tests/auth_spec.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
+const backend = 'localhost:3001';
+const testUsername = 'test';
+const testPassword = 'test';
+// Must be running the auth server for all tests.
```
Comment (caoimhegall) [inline frontend/cypress/integration/tests/auth_spec.js]: This doesn't need to be a Cypress test. Cypress is used to test frontend elements like clicking on pages or components, typing in fields etc. What you wrote here should be a Jest test instead.

File: backend/routes/userRoutes.js [hunk 8]
```diff
@@ -43,47 +79,49 @@ router.get('/users/:userId/email', function(req, res) {
...
-                    userData.favoriteIds.includes(serviceDoc.serviceId);
-                return serviceObj;
-              });
-              res.send(servicesWithFavorites);
-            }
-          });
-    }
-  });
-});
+router.get('/users/services',
```
Comment (adrien-chaussabel) [inline backend/routes/userRoutes.js]: Why did you remove accessing a single user through :userId?

File: backend/routes/userRoutes.js [hunk 9]
```diff
@@ -43,47 +79,49 @@ router.get('/users/:userId/email', function(req, res) {
...
+          }
+        }
+    );
+  }
+);
 
 // Route for adding a subscribed service to a specific user
-router.post('/users/:userId/services', function(req, res) {
-  usersModel.findOneAndUpdate({userId: req.params.userId},
+router.post('/users/services',
```
Comment (adrien-chaussabel) [inline backend/routes/userRoutes.js:104]: Again here, confused on why :userId was removed from route

File: backend/routes/authRoutes.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+})
+
+// Route to login ~ recieve accessToken, refreshToken, and loggedIn cookie
+router.post('/login',  (req, res) => {
+  usersModel.findOne({username: req.body.username})
+    .then((user) => {
+      if (!user) {
+        res.status(404).json({msg: "could not find user"});
+      }
+      bcrypt.compare(req.body.password, user.password).then(match =>
```
Comment (ncpalmie) [inline backend/routes/authRoutes.js]: When a username is entered that doesn't exist in the database, this function doesn't exit before the bcrypt compare causing errors in backend.

**Review by ncpalmie — COMMENTED** [review]
Overall code looks great and looks good to merge. I was able to successfully login and register a user. The only issue I noticed was with the login route mentioned below.

File: backend/routes/authRoutes.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
+require('dotenv').config();
```
Comment (adrien-chaussabel) [inline backend/routes/authRoutes.js:1]: There are a host of ESLint errors on this page and others in the PR.

File: backend/routes/authRoutes.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+const express = require('express');
```
Comment (adrien-chaussabel) [inline backend/routes/authRoutes.js:2]: Organizing routes is a good idea

File: backend/routes/authRoutes.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+const cookieParser = require('cookie-parser');
+
+// Mongoose models
+const usersModel = require('../models/users');
+let refreshTokens = []; // should use a mongoose model in the future
+
+router.use(cookieParser());
+// duration of access/refresh cookies in seconds
+let accessTime = 60;
+let refreshTime = 3600;
```
Comment (adrien-chaussabel) [inline backend/routes/authRoutes.js]: Good to store as constants

File: frontend/src/axiosConfig.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+import axios from 'axios';
+axios.defaults.withCredentials = true;
+/**
+ *
```
Comment (adrien-chaussabel) [inline frontend/src/axiosConfig.js]: Explain what this is for

File: backend/routes/authRoutes.js (hunk 11 above)
Comment (caoimhegall) [inline backend/routes/authRoutes.js:1]: Make sure you have the eslint extension enabled in VSCode. There are a lot of eslint errors in your code.

File: frontend/cypress/integration/tests/auth_spec.js (hunk 7 above)
Comment (caoimhegall) [reply frontend/cypress/integration/tests/auth_spec.js]: You could add a cypress test for logging in and registering, just use cy.get and cy.click to interact with frontend elements

**Review by adrien-chaussabel — CHANGES_REQUESTED** [review]
Overall the JWT implementation is solid, but has some issues as discussed in the pre-review meeting. There's also a lot of ESLint issues that should be resolved before merging this PR. I'm also not sure if tests should be done through cypress, as they seem to be mostly backend testing that can be done through Jest.

**Review by caoimhegall — COMMENTED** [review]
Login and register look good. Need to fix the "secure" issue we talked about during the meeting. Also install eslint and fix those issues.
